codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,469 @@
+        decoder.frameListener(new FrameListener());
+        http2Handler = new InternalHttp2ConnectionHandler(decoder, encoder, new Http2Settings());
+        http2Handler.connection().addListener(new ConnectionListener());
+        streamInfoKey = http2Handler.connection().newKey();
+    }
+
+    Http2ConnectionHandler connectionHandler() {
+        return http2Handler;
+    }
+
+    /**
+     * Save context and load any dependencies.
+     */
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        ctx.pipeline().addBefore(ctx.executor(), ctx.name(), null, http2Handler);
+        this.http2HandlerCtx = ctx.pipeline().context(http2Handler);
+    }
+
+    /**
+     * Clean up any dependencies.
+     */
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+      ctx.pipeline().remove(http2Handler);
+    }
+
+    /**
+     * Handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via
+     * HTTP/2 on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
+     */
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!(evt instanceof UpgradeEvent)) {
+            super.userEventTriggered(ctx, evt);
+            return;
+        }
+
+        UpgradeEvent upgrade = (UpgradeEvent) evt;
+        ctx.fireUserEventTriggered(upgrade.retain());
+        try {
+            Http2Stream stream = http2Handler.connection().stream(Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
+            // TODO: improve handler/stream lifecycle so that stream isn't active before handler added.
+            // The stream was already made active, but ctx may have been null so it wasn't initialized.
+            new ConnectionListener().onStreamActive(stream);
+            upgrade.upgradeRequest().headers().setInt(
+                HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
+            new InboundHttpToHttp2Adapter(http2Handler.connection(), http2Handler.decoder().frameListener())
+                .channelRead(ctx, upgrade.upgradeRequest().retain());
+        } finally {
+            upgrade.release();
+        }
+    }
+
+    /**
+     * Processes all {@link Http2Frame}s. {@link Http2StreamFrame}s may only originate in child
+     * streams.
+     */
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+            throws Exception {
+        if (!(msg instanceof Http2Frame)) {
+            super.write(ctx, msg, promise);
+            return;
+        }
+        try {
+            if (msg instanceof Http2StreamFrame) {
+                Object streamObject = ((Http2StreamFrame) msg).stream();
+                int streamId = ((Http2StreamChannel) streamObject).stream.id();
+                if (msg instanceof Http2DataFrame) {
```

Comment:
> but I like putting frames in control of writing them selves more

Does any codec currently do that?

> Less chance of having to update the "core" code when updating/adding frames.

How does a `writeFrame` method solve reading frames? `writeFrame` is also depending on `Http2FrameWriter`, which... doesn't really make the situation any better. I could maybe understand a `ByteBuf encode()` method, but without also being able to have custom logic for reading frames it doesn't seem to solve much.

Requiring the write logic to be in the frame also makes implementing your own version of a built-in frame seem pretty limited because if that solution was complete (say, with flow control support, so splitting DATA frames, and HEADERS encoding themselves) then the effort to reimplement a concrete version of the interface becomes prohibitive.

Arbitrary frame support is non-trivial, _especially_ if trying to use it for built-in frame types since they influence each other. The only built-in frame type I see that could actually be implemented without changing the "core" is PING, and that is trivial to implement. Even if it is to just support a new flag, the only flag I see that is "plug and play" is PADDED. So while I see the point of being able to make your own frames, assuming that you can make a custom, non-trivial frame without modifying the core seems unlikely.

Also, we are making use of the pipeline again. Soon, applications could add their own handlers that could convert there own custom messages to ByteBufs and send them.